modules/configuration/ajax/process.php

```diff
@@ -30,65 +32,112 @@
-                   exit();
+            if (! noDuplicateInDropdown($key, $value)) {
+                // Don't alter the table if the same key was passed twice.
+                continue;
+            }
+            // Get all the IDs in Config with a relation to an entry in the
+            // ConfigSettings table that has the web_path data type.
+            $pathIDs = getPathIDs('Config');
+            if (in_array($key, $pathIDs)) {
+                // If updating a path config setting, make sure that the
+                // value is a valid path. This helps to prevent malicious
+                // input.
+                if (!validPath($value)) {
+                    $err = 'Directory `'
+                        . htmlspecialchars($value, ENT_QUOTES)
+                        . '` is invalid';
+                    // Set response code and display error message.
+                    displayError(400, $err);
+                    return;
+                }
             }
+            // Update the config setting to the new value.
+            $DB->update(
+                'Config',
+                array('Value' => $value),
+                array('ID' => $key)
+            );
         }
-    } else { //add new or remove
-        $keySplit   = explode("-", $key);
-        $valueSplit = explode("-", $value);
-        if ($keySplit[0] == 'add') {
-            if ($value !== "") {
-                if (countDuplicate($keySplit[1], $value) == '0') {
-                    $DB->insert(
-                        'Config',
-                        array(
-                         'ConfigID' => $keySplit[1],
-                         'Value'    => $value,
-                        )
-                    );
-                } else {
-                        header("HTTP/1.1 303 Duplicate value");
-                        exit();
+    } else {
+        // This branch is executed when the key is prefixed with the string
+        // 'add' or 'remove' (i.e. the full key is not numeric and triggers this
+        // else branch).
+        // An example $key is 'add-38' which means "Add an entry in the Config
+        // table using foriegn key 38 from the ConfigSettings table." The
+        // number refers to ConfigSettings.ID which is different from
+        // Config.ID.
+        // This is different from the above is_numeric case; this makes use of
+        // Config.ID, not Config.ConfigID (which is a FK to ConfigSettings.ID).
+        // The Config table is the one that will be modified here.
+        $keySplit         = explode("-", $key); // e.g. 'add-10' or 'remove-49'
+        $action           = $keySplit[0];
+        $ConfigSettingsID = $keySplit[1];
+        if ($action == 'add') {
+            // This branch adds a new entry to the Config table.
+            if ($value === "") {
+                continue;
+            }
+            if (isDuplicate($ConfigSettingsID, $value)) {
+                header("HTTP/1.1 303 Duplicate value");
```

Comment:
HTTP/1.1 defines A 303 response code is "See Other", not "Duplicate Value"

Reply:
OK. I didn't write this. Should we change this to Bad Request?

Reply:
to be honest I'm not entirely sure what the "duplicate" check is trying to prevent and how it relates to settings which have AllowMultiple set.. but if anything,  yes, `400 Bad Request` is at least a valid HTTP response code.

Reply:
Yeah me neither.